src/VisualStudio/LiveShare/Impl/Client/RemoteLanguageServiceWorkspace.cs

```diff
@@ -94,100 +113,134 @@ void IRunningDocumentTableEventListener.OnRenameDocument(string newMoniker, stri
             // Handled by Add/Remove.
         }
 
-        public async Task SetSession(CollaborationSession session)
+        public async Task SetSessionAsync(CollaborationSession session)
         {
             _session = session;
-            var roots = await session.ListRootsAsync(CancellationToken.None).ConfigureAwait(false);
-            if (roots.Length > 0)
-            {
-                _remoteRootPath = session.ConvertSharedUriToLocalPath(roots[0]);
-                _remoteRootPath = _remoteRootPath.Substring(0, _remoteRootPath.Length - 1);
-                var lastSlash = _remoteRootPath.LastIndexOf('\\');
-                _externalPath = _remoteRootPath.Substring(0, lastSlash + 1);
-                _externalPath += "~external";
-            }
-            IsRemoteSession = true;
+
+            StartSolutionCrawler();
+
+            // Get the initial workspace roots and update any files that have been opened.
+            await UpdatePathsToRemoteFilesAsync(session).ConfigureAwait(false);
+
             session.RemoteServicesChanged += (object sender, RemoteServicesChangedEventArgs e) =>
             {
                 _remoteDiagnosticListTable.UpdateWorkspaceDiagnosticsPresent(_session.RemoteServiceNames.Contains("workspaceDiagnostics"));
             };
         }
 
-        public void EndSession()
-        {
-            IsRemoteSession = false;
-        }
-
-        public void Init()
-        {
-            StartSolutionCrawler();
-        }
+        public string? GetRemoteExternalRoot(string filePath)
+            => _registeredExternalPaths.SingleOrDefault(externalPath => filePath.StartsWith(externalPath));
 
-        /// <inheritdoc />
-        public override bool CanOpenDocuments => true;
+        public string? GetRemoteWorkspaceRoot(string filePath)
+            => _remoteWorkspaceRootPaths.SingleOrDefault(remoteWorkspaceRoot => filePath.StartsWith(remoteWorkspaceRoot));
 
-        /// <inheritdoc />
-        public void OnManagedProjectAdded(ProjectInfo projectInfo)
+        /// <summary>
+        /// Event that gets triggered whenever the active workspace changes.  If we're in a live share session
+        /// this means that the remote workpace roots have also changed and need to be updated.
+        /// This will not be called concurrently.
+        /// </summary>
+        private async Task OnActiveWorkspaceChangedAsync(object sender, EventArgs args)
         {
-            //Notify AnyCode Roslyn workspace a project is being added.
-            OnProjectAdded(projectInfo);
+            if (IsRemoteSession)
+            {
+                await UpdatePathsToRemoteFilesAsync(_session!).ConfigureAwait(false);
```

Comment:
Can this session still be closed between here and the "IsRemoteSession" check?

Reply:
Or, just replace the IsRemoteSession check with session != null so the compiler knows this? It sounds like the compiler is also adding a way we might be able to annotate IsRemoteSession so the compiler will know the connection.

Reply:
There's nothing preventing that from happening.  I added the check under the gate to try and address the case where the session ends in between.  Since the end session also takes the gate it I think it should be safe